From 95bc50e18b33b8373193294bc39cbdf066d27158 Mon Sep 17 00:00:00 2001
From: bo liu <bliu0@pek-lpgbuild5.(none)>
Date: Wed, 24 Feb 2010 01:33:11 -0800
Subject: [PATCH 1/2] arm_versatile_926ejs: add default nor flash mtd parts

Create default MTD partition tables for versatile and fix the problem
that without partition the integrator map driver will fail to register
mtd device.

Signed-off-by: Tony Liu <Bo.Liu@windriver.com>
---
 arch/arm/mach-versatile/core.c      |   33 +++++++++++++++++++++++++++++++++
 drivers/mtd/maps/integrator-flash.c |   13 ++++++++-----
 2 files changed, 41 insertions(+), 5 deletions(-)

diff --git a/arch/arm/mach-versatile/core.c b/arch/arm/mach-versatile/core.c
index e13be7c..42f8bf0 100644
--- a/arch/arm/mach-versatile/core.c
+++ b/arch/arm/mach-versatile/core.c
@@ -22,6 +22,8 @@
 #include <linux/device.h>
 #include <linux/dma-mapping.h>
 #include <linux/platform_device.h>
+#include <linux/mtd/mtd.h>
+#include <linux/mtd/partitions.h>
 #include <linux/sysdev.h>
 #include <linux/interrupt.h>
 #include <linux/amba/bus.h>
@@ -284,12 +286,43 @@ static void versatile_flash_set_vpp(int on)
 	__raw_writel(val, VERSATILE_FLASHCTRL);
 }
 
+static struct mtd_partition versatile_partitions[] = {
+	{ /* bootloader and params */
+		.name           = "bootloader",
+		.offset         = 0,
+		.size           = SZ_256K +  SZ_128K,
+		.mask_flags     = MTD_WRITEABLE,        /* force read-only */
+	}, {
+		.name           = "kernel",
+		.offset         = MTDPART_OFS_APPEND,
+		.size           = (0x400000 - 0x60000),
+		.mask_flags     = 0,
+	}, {
+		.name           = "jffs2",
+		.offset         = MTDPART_OFS_APPEND,
+		.size           = 0x700000,
+		.mask_flags     = 0,
+	}, {
+		.name           = "cramfs",
+		.offset         = MTDPART_OFS_APPEND,
+		.size           = 0x200000,
+		.mask_flags     = 0,
+	}, {
+		.name           = "user data",
+		.offset         = MTDPART_OFS_APPEND,
+		.size           = MTDPART_SIZ_FULL,
+		.mask_flags     = 0,
+	}
+};
+
 static struct flash_platform_data versatile_flash_data = {
 	.map_name		= "cfi_probe",
 	.width			= 4,
 	.init			= versatile_flash_init,
 	.exit			= versatile_flash_exit,
 	.set_vpp		= versatile_flash_set_vpp,
+	.parts                  = versatile_partitions,
+	.nr_parts               = ARRAY_SIZE(versatile_partitions),
 };
 
 static struct resource versatile_flash_resource = {
diff --git a/drivers/mtd/maps/integrator-flash.c b/drivers/mtd/maps/integrator-flash.c
index 2aac41b..11b9229 100644
--- a/drivers/mtd/maps/integrator-flash.c
+++ b/drivers/mtd/maps/integrator-flash.c
@@ -225,13 +225,16 @@ static int armflash_probe(struct platform_device *dev)
 	if (err < 0)
 		goto cleanup;
 
+#ifdef CONFIG_MTD_PARTITIONS
 	err = parse_mtd_partitions(info->mtd, probes, &info->parts, 0);
-	if (err > 0) {
+	if (err > 0)
 		err = add_mtd_partitions(info->mtd, info->parts, err);
-		if (err)
-			printk(KERN_ERR
-			       "mtd partition registration failed: %d\n", err);
-	}
+	else if (err <= 0 && plat->parts)
+		err = add_mtd_partitions(info->mtd, plat->parts,
+			plat->nr_parts);
+	else
+#endif
+		err = add_mtd_device(info->mtd);
 
 	if (err == 0) {
 		platform_set_drvdata(dev, info);
-- 
1.6.5.2

